### PR TITLE
Edit rescale_intensity to return original image instead of nans

### PR DESCRIPTION
**Pull request by marshuang80**
## Description
This PR was originally addressing issue stated in #3773, which requests CLAHE to return numpy of zeros instead of nans when the input is np.zeros(n,n). 

Further examining the code indicates that the problem stems from the rescale_intensity() function in equalize_adapthist(), which returns np.array of nans if all values in the input image are the same. This PR edits rescale_intensity() to return the original image if the max and min of the image is the same. 

## Checklist

- [X ] [Docstrings for all functions](https://github.com/numpy/numpy/blob/master/doc/example.py)
- [X ] Clean style in [the spirit of PEP8](https://www.python.org/dev/peps/pep-0008/)

## For reviewers

- [ ] Check that the PR title is short, concise, and will make sense 1 year
  later.
- [ ] Check that new functions are imported in corresponding `__init__.py`.
- [ ] Check that new features, API changes, and deprecations are mentioned in
      `doc/release/release_dev.rst`.
- [ ] Consider backporting the PR with `@meeseeksdev backport to v0.14.x`

### CODE REVIEW

File: skimage/exposure/exposure.py [hunk 1]
```diff
@@ -341,6 +341,9 @@ def rescale_intensity(image, in_range='image', out_range='dtype'):
     array([  0,  63, 127], dtype=int8)
 
     """
+    if image.min() == image.max():
```
Comment (hmaarrfk) [inline skimage/exposure/exposure.py]: I feel like you should use `imin` and `imax` computed below. The check should likely be done just before line 354 is computed. Where by you compare the denominator to `0` to check if you are dividing by zero.

Comment (marshuang80) [reply skimage/exposure/exposure.py]: That makes sense. Just to clarify, we should return the clipped image instead of the original image right?

Comment (hmaarrfk) [reply skimage/exposure/exposure.py]: Yes i believe so. The output would therefore be `omin` after all the math is performed.

I generally think we shouldn't spend too much time special casing edge cases like this, but not returning `nan` makes sense.

Comment (marshuang80) [reply skimage/exposure/exposure.py]: Sounds good, I've updated my PR anyways. 

There is a failed test from Travis but is seems unrelated to the fix.

**Comment by hmaarrfk** [comment]
I think i mean that we require a *new* test. One that **would** have failed before your change, but passes now because of your change.

File: skimage/exposure/exposure.py (hunk 1 above)
Comment (hmaarrfk) [reply skimage/exposure/exposure.py]: These two lines should be removed right?

File: skimage/exposure/exposure.py [hunk 2]
```diff
@@ -348,6 +348,8 @@ def rescale_intensity(image, in_range='image', out_range='dtype'):
 
     image = np.clip(image, imin, imax)
 
+    if imin == imax:
+        return image
```
Comment (hmaarrfk) [inline skimage/exposure/exposure.py]: I would really like it if the only thing that was avoided here was the division by zero.

The code path should go through the same trajectory multiplying by `(omax - omin)` and adding `omin`.

Sound riddiculous, but sometimes not going through that code path can hide nasty bugs due to returning a view of the same image, as opposed to a new array like the code below does.

File: skimage/exposure/exposure.py [hunk 3]
```diff
@@ -348,6 +348,8 @@ def rescale_intensity(image, in_range='image', out_range='dtype'):
...
     image = (image - imin) / float(imax - imin)
     return np.array(image * (omax - omin) + omin, dtype=dtype)
```
Comment (hmaarrfk) [inline skimage/exposure/exposure.py]: can you please make this `asarray` instead of `np.array`, it may avoid a copy if the image is already a float.

Comment (hmaarrfk) [reply skimage/exposure/exposure.py]: Sorry. I was referring to line 354, the final line.

File: skimage/exposure/exposure.py (hunk 2 above)
Comment (marshuang80) [reply skimage/exposure/exposure.py]: Ah I see. So should I do the following check instead of returning an original copy? 
```if imin` != imax: image = (image - imin) / float(imax - imin)```

File: skimage/exposure/exposure.py (hunk 3 above)
Comment (marshuang80) [reply skimage/exposure/exposure.py]: Okay, sounds good. Should I still return the original image if max == min? Or should I just skip the division on line 353 if max != min?

Comment (hmaarrfk) [reply skimage/exposure/exposure.py]: Just skip the division part.

Comment (hmaarrfk) [reply skimage/exposure/exposure.py]: I want to return omin if the image is a constant. Get it?

Comment (marshuang80) [reply skimage/exposure/exposure.py]: Yep, got it

**Comment by marshuang80** [comment]
Sorry can you clarify on what I am testing?

**Comment by hmaarrfk** [comment]
I think that is more nitpicking.

But more that zeros is generally an edge case. I tend not to use it when testing for general "constants"

**Comment by marshuang80** [comment]
Okay sounds good, I just made another test with non-zero constants

File: skimage/exposure/tests/test_exposure.py [hunk 4]
```diff
@@ -267,6 +267,18 @@ def test_rescale_uint14_limits():
     assert_array_almost_equal(out, [0, uint14_max])
 
 
+def test_rescale_all_zeros():
+    image = np.zeros((2, 2), dtype=np.uint8)
+    out = exposure.rescale_intensity(image)
+    assert ~np.isnan(out).all()
```
Comment (hmaarrfk) [inline skimage/exposure/tests/test_exposure.py]: We know what the output shouldn't be. Do we know what the output should be?

Comment (marshuang80) [reply skimage/exposure/tests/test_exposure.py]: Yep, it should be the same

Comment (hmaarrfk) [reply skimage/exposure/tests/test_exposure.py]: it is probably because of the changes I requested. I think due to the logic that you have now, omin and omax are the same, and thus the image will be rescaled to zero. Not sure what the right behavior here is.
